File: CS/WpfApplication2/Converters.cs
using System;
using System.Collections;
using System.Globalization;
using System.Windows.Data;

namespace WpfApplication2
{
	public class CollectionToIsCheckedConverter : IMultiValueConverter
	{
		private ViewModel dataContext;

		public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture)
		{
			ICollection collection = values[0] as ICollection;
			dataContext = values[1] as ViewModel;

			if ( collection == null || dataContext == null )
				throw new NotSupportedException();

			if ( collection.Count == 0 )
				return false;

			if ( collection.Count == dataContext.List.Count )
				return true;

			return null;
		}

		public object[] ConvertBack(object value, Type[] targetType, object parameter, CultureInfo culture)
		{
			throw new NotSupportedException();
		}
	}

	public class CollectionToIsThreeStateConverter : IMultiValueConverter
	{
		public object Convert(object[] values, Type targetType, object parameter, CultureInfo culture)
		{
			ICollection collection = values[0] as ICollection;
			ViewModel dataContext = values[1] as ViewModel;

			if ( collection == null || dataContext == null )
				throw new NotSupportedException();

			if ( collection.Count == 0 || collection.Count == dataContext.List.Count )
				return false;

			return true;
		}

		public object[] ConvertBack(object value, Type[] targetType, object parameter, CultureInfo culture)
		{
			throw new NotSupportedException();
		}
	}
}

File: VB/WpfApplication2/Converters.vb
Imports System
Imports System.Collections
Imports System.Globalization
Imports System.Windows.Data

Namespace WpfApplication2
	Public Class CollectionToIsCheckedConverter
		Implements IMultiValueConverter

		Private dataContext As ViewModel

		Public Function Convert(ByVal values() As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IMultiValueConverter.Convert
			Dim collection As ICollection = TryCast(values(0), ICollection)
			dataContext = TryCast(values(1), ViewModel)

			If collection Is Nothing OrElse dataContext Is Nothing Then
				Throw New NotSupportedException()
			End If

			If collection.Count = 0 Then
				Return False
			End If

			If collection.Count = dataContext.List.Count Then
				Return True
			End If

			Return Nothing
		End Function

		Public Function ConvertBack(ByVal value As Object, ByVal targetType() As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
			Throw New NotSupportedException()
		End Function
	End Class

	Public Class CollectionToIsThreeStateConverter
		Implements IMultiValueConverter

		Public Function Convert(ByVal values() As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object Implements IMultiValueConverter.Convert
			Dim collection As ICollection = TryCast(values(0), ICollection)
			Dim dataContext As ViewModel = TryCast(values(1), ViewModel)

			If collection Is Nothing OrElse dataContext Is Nothing Then
				Throw New NotSupportedException()
			End If

			If collection.Count = 0 OrElse collection.Count = dataContext.List.Count Then
				Return False
			End If

			Return True
		End Function

		Public Function ConvertBack(ByVal value As Object, ByVal targetType() As Type, ByVal parameter As Object, ByVal culture As CultureInfo) As Object() Implements IMultiValueConverter.ConvertBack
			Throw New NotSupportedException()
		End Function
	End Class
End Namespace

